numpy/lib/format.py

```diff
@@ -541,7 +543,8 @@ def write_array(fp, array, version=None, pickle_kwargs=None):
     Raises
     ------
     ValueError
-        If the array cannot be persisted.
+        If the array cannot be persisted. This includes the case of
+        allow_pickle=True and array being an object array.
```

Comment:
From the code, it looks like `allow_pickle=False` and object array is what raises the error.